src/super_gradients/training/models/segmentation_models/segformer.py

```diff
@@ -372,12 +386,58 @@ def _remove_auxiliary_heads(self):
     def replace_head(self, new_num_classes: int, new_decoder_embed_dim: int):
         self.decode_head = SegFormerHead(encoder_dims=self.encoder_embed_dims, embed_dim=new_decoder_embed_dim, num_classes=new_num_classes)
 
-    def forward(self, x: torch.Tensor) -> torch.Tensor:
+    def _forward(self, x: torch.Tensor) -> torch.Tensor:
         features = self._backbone(x)
         out = self.decode_head(features)
         out = F.interpolate(out, size=x.shape[2:], mode="bilinear", align_corners=False)
         return out
 
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        if self.is_SWI:
+            return self.sliding_window(x, self.num_classes, self.sliding_window_stride, self.sliding_window_size)
+        else:
+            return self._forward(x)
+
+    def sliding_window(self, img: torch.Tensor, num_classes: int, stride: Tuple[int, int], crop_size: Tuple[int, int]) -> torch.Tensor:
+        """
+        Inference by sliding-window with overlap.
+
+        If h_crop > h_img or w_crop > w_img, the small patch will be used to
+        decode without padding.
+        """
+
+        h_stride, w_stride = stride
+        h_crop, w_crop = crop_size
+        batch_size, _, h_img, w_img = img.size()
+
+        h_grids = max(h_img - h_crop + h_stride - 1, 0) // h_stride + 1
+        w_grids = max(w_img - w_crop + w_stride - 1, 0) // w_stride + 1
+
+        preds = torch.zeros((batch_size, num_classes, h_img, w_img), device=img.device)
+        count_mat = torch.zeros((batch_size, 1, h_img, w_img), device=img.device)
+
+        for h_idx in range(h_grids):
+            for w_idx in range(w_grids):
+                y1 = h_idx * h_stride
+                x1 = w_idx * w_stride
+                y2 = min(y1 + h_crop, h_img)
+                x2 = min(x1 + w_crop, w_img)
+                y1 = max(y2 - h_crop, 0)
+                x1 = max(x2 - w_crop, 0)
+                crop_img = img[:, :, y1:y2, x1:x2]
+                crop_logits = self._forward(crop_img)
+
+                if isinstance(crop_logits, torch.Tensor):
+                    crop_logits = (crop_logits,)
+
+                crop_logits = crop_logits[0]
+
+                preds += F.pad(crop_logits, pad=(int(x1), int(preds.shape[3] - x2), int(y1), int(preds.shape[2] - y2)))
+
+                count_mat[:, :, y1:y2, x1:x2] += 1
```

Comment:
Please add unit tests covering the different scenarios in SWI, making sure we always cover the entire image (i,e count mat should never have a 0 entry)

Reply:
That may require refactoring the sliding inference logic to an external method to check whether accumulator and count_mat contains expected values, but I agree that should be tested thoroughly.

Reply:
@shaydeci how should I proceed?

Reply:
I'm not Shay, but my 5 cents on how to start:

1) Extract `sliding_window_inference(model, input, tile_size, tile_step, ...)` method that would do just the sliding inference of user-provided model. That would also enable us to use this method in other models.

2) Add unit tests to check the correctness of the function for following cases:

* Input size is smaller than tile_size: `input_size=(512,512`, `tile_size=(640,640)`
* Input size is equal to tile size
* Input size is greater than tile size `input_size=(513,513)`, `tile_size=(512,512)`
* For each of these cases you may also want to check: 
* tile_step is equal to tile_size
* tile_step is larger than tile size (This should raise an Exception)
* tile_step is less than tile_size (tile_size=512, tile_step=384)
* Also check that function works correctly with odd-sized inputs (input_size=13, tile_step=3, tile_size=2)

3) How to check correctness
Use dummy model that returns an input. So the `sliding_window_inference` should then reconstruct the same input (torch.randn(input_size))

Reply:
I think @BloodAxe covered it pretty much here ^

Reply:
in tests/unit_tests as the other unit tests.
Dont forget to add it to the test suite.

Reply:
Fixed on e9a6e9f725e8730efe9130f78b2aa604ed253732, 61ea395063f51661a3f6c8e26e0ce050bc73ffd9, 1475d135331fb8ef7ade9723b4e1a95313ada41a and 0ec8a9cf3c0ae469c9b02018cfa477177caaedb3.